[16.0][ADD] internal_stock_quant_package

glitchov opened the pull request:
Defines internal package used in picking process and are emptied (or not upon config) at 'put in pack' or when the picking is validated.

glitchov committed:
[16.0][ADD] internal_stock_quant_package

lmignon commented on internal_stock_quant_package/models/stock_picking_type.py:
```diff
@@ -0,0 +1,32 @@
+    @api.model
+    @tools.ormcache("picking_type_id", "carrier_id")
+    def _empty_internal_package_on_transfer(self, picking_type_id, carrier_id):
+        picking_type = self.browse(picking_type_id)
```
@glitchov We should explain the purpose of this method....

```In this method we check if the internal package must be emptied for the given picking type and carrier
To know if internal packages must be emptied:
1. We check if the flag is set on the picking type
2. We lookup if a specific configuration exists for the given partner
3. If a specific configuration exists we return the value of this configuration
4. Otherwise we return the value on the picking type.
```

lmignon commented on internal_stock_quant_package/models/stock_picking.py:
```diff
@@ -0,0 +1,78 @@
+        compute="_compute_empty_internal_package_on_transfer",
+    )
+
+    def _get_out_picking(self):
+        # this method is coupled with sale.order _compute_picking_ids, so any override
+        # there should have an equivalent override of this method
+        self.ensure_one()
+        domain = [
+            ("group_id", "=", self.group_id.id),
+            ("picking_type_id.code", "=", "outgoing"),
+            ("state", "not in", ["cancel", "done"]),
+        ]
+        return self.search(domain, limit=1)  # what should we do if we got more than 1?
+
+    def _get_carrier(self):
+        self.ensure_one()
+        if self.picking_type_id.code != "outgoing":
+            carrier = self._get_out_picking().carrier_id
+        else:
+            carrier = self.carrier_id
+        return carrier
```
@glitchov You should depends on delivery_procurement_group_carrier (https://github.com/OCA/stock-logistics-workflow/pull/1158)
This addon add the carrier_id on the procurement group. With this addon we avoid a seach to get the carrier_id....

``` python
return self.group_id.carrier_id
```

lmignon replied:
@glitchov you should remove the two methods `_get_out_picking` and `_get_carrier` and adapt the compute method to use the `carrier_id` or `group_id.carrier_id` directly

```python
 @api.depends("picking_type_id.empty_internal_package_on_transfer", "carrier_id", "group_id.carrier_id")
    def _compute_empty_internal_package_on_transfer(self):
        for record in self:
            carrier = record.carrier_id or record.group_id.carrier_id
            carrier_id = carrier.id
```

lmignon commented on internal_stock_quant_package/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,6 @@
+This module allows to declare internal stock quant package.
+
+An internal stock quant package will never leave the warehouse. It's
+a package used in the picking process and is emptied when a 'put
+in pack' operation occurs or when the picking is validated (except
+if configured to not do it on the picking type).
```
@glitchov We should be more precise
 ```rst
Sometimes, when an operator is picking, he need to put the product in
internal packages placed on his trolley that will be emptied later.

Two kinds of operations can lead to the emptying of the internal packages:
* when product from the internal packages will be 'put in pack' at the pack station
  (in a pick / pack / ship scenario)
* when a carrier will load his truck with the products from the internal packages
  (in a pick / ship scenario)

This modules extends the stock module to add the concept of internal stock
quant package and therefore allows you to manage this kind of operational need.
It ensures that the internal stock quant packages are emptied when required
depending on the picking type configuration.
```

We should also add a USAGE.rst file with

```rst
As this addon rely on the concept of "internal" packages. If you want to
use packages into your picking operations, you need first to activate the
package functionality in the stock settings (see the "Operations" section).

Then, you need to create packages and set them as internal. This is done
by going to Inventory > Products > Packages and clicking on the "Create".
(Don't forget to tick the "Internal use" box).

By default, when you put your products into an internal package when processing
a picking, once the picking is done, the package is automatically emptied.
You can change this behavior at 2 levels:
1. At the picking type level: go to "Inventory > Configuration > Operation
  Types" and edit the picking type you want to change. Then, un tick the "Empty
  Internal Package On Transfer" box. (By default internal packages are always
  emptied when the picking is done).
2 At the picking type level for a specific carrier: go to "Inventory >
  Configuration > Operation Types" and edit the picking type you want to change.
  Then, add or remove lines in the "Stock Internal Package Config Line" table.
  You can add a line for a specific carrier and tick /un tick the "Empty" box.

To know if internal packages must be emptied or not for a given picking, the
system will first check if a configuration line exists on the picking type for
the carrier of the picking. If a line exists, the system will use the value
of the "Empty" box. If no line exists, the system will use the value of the
"Empty Internal Package On Transfer" box of the picking type.

```

lmignon commented on internal_stock_quant_package/readme/USAGE.rst:
```diff
@@ -0,0 +1,27 @@
+a picking, once the picking is done, the package is automatically emptied.
+You can change this behavior at 2 levels:
+
+1. At the picking type level: go to "Inventory > Configuration > Operation
+Types" and edit the picking type you want to change. Then, untick the "Empty
+Internal Package On Transfer" box. (By default internal packages are always
+emptied when the picking is done).
+
+2. At the picking type level for a specific carrier: go to "Inventory >
+Configuration > Operation Types" and edit the picking type you want to change.
+Then, add or remove lines in the "Stock Internal Package Config Line" table.
+You can add a line for a specific carrier and tick/untick the "Empty" box.
```
```suggestion
1. At the picking type level: go to "Inventory > Configuration > Operation
  Types" and edit the picking type you want to change. Then, untick the "Empty
  Internal Package On Transfer" box. (By default internal packages are always
  emptied when the picking is done).
2. At the picking type level for a specific carrier: go to "Inventory >
  Configuration > Operation Types" and edit the picking type you want to change.
  Then, add or remove lines in the "Stock Internal Package Config Line" table.
  You can add a line for a specific carrier and tick/untick the "Empty" box.
```

lmignon commented on requirements-test.txt:
```diff
@@ -0,0 +1,2 @@
+# https://github.com/OCA/stock-logistics-workflow/pull/1158
+odoo-addon-delivery-procurement-group-carrier @ git+https://github.com/acsone/stock-logistics-workflow.git@16.0-mig-delivery_procurement_group_carrier-dro#subdirectory=setup/delivery_procurement_group_carrier
```
@glitchov the filename should be `test-requirements.txt' Moreover, this file should be added into a separate commit with "DO NOT MERGE" as message

lmignon reviewed (approved):
LGTM (Code review + functional tests)

rousseldenis committed:
[IMP] internal_stock_quant_package: Don't call unpack if not necessary

As Odoo unpack() is calling the _merge_quants() function which can be bulky,
don't call it if not necessary

rousseldenis commented:
/ocabot merge nobump
